tests/e2e/improved_csi_idempotency.go

```diff
@@ -0,0 +1,537 @@
+		newSize.Add(resource.MustParse("4Gi"))
+		framework.Logf("currentPvcSize %v, newSize %v", currentPvcSize, newSize)
+		claims, err := expandPVCSize(claim, newSize, client)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(claims).NotTo(gomega.BeNil())
+	}
+
+	if serviceName == "CSI" {
+		ginkgo.By("Stopping CSI driver")
+		collectPodLogs(ctx, client, csiSystemNamespace)
+		err = updateDeploymentReplicawithWait(client, 0, vSphereCSIControllerPodNamePrefix, csiSystemNamespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		framework.Logf("Starting CSI driver")
+		err = updateDeploymentReplicawithWait(client, 1, vSphereCSIControllerPodNamePrefix, csiSystemNamespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		// Wait for the CSI Pods to be up and Running
+		list_of_pods, err := fpod.GetPodsInNamespace(client, csiSystemNamespace, ignoreLabels)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		num_csi_pods := len(list_of_pods)
+		fpod.WaitForPodsRunningReady(client, csiSystemNamespace, int32(num_csi_pods), 0, pollTimeout, ignoreLabels)
+
+		ginkgo.By(fmt.Sprintf("Sleeping for %v seconds to allow full sync finish", fullSyncWaitTime))
+		time.Sleep(time.Duration(fullSyncWaitTime) * time.Second)
+	} else {
+		ginkgo.By(fmt.Sprintf("Stopping %v on the vCenter host", serviceName))
+		vcAddress := e2eVSphere.Config.Global.VCenterHostname + ":" + sshdPort
+		err = invokeVCenterServiceControl(stopOperation, serviceName, vcAddress)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = true
+		err = waitVCenterServiceToBeInState(serviceName, vcAddress, stopMessage)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		defer func() {
+			if isServiceStopped {
+				ginkgo.By(fmt.Sprintf("Starting %v on the vCenter host", serviceName))
+				err = invokeVCenterServiceControl(startOperation, serviceName, vcAddress)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				err = waitVCenterServiceToBeInState(serviceName, vcAddress, startMessage)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			}
+		}()
+
+		ginkgo.By(fmt.Sprintln("Sleeping for 5+1 min for default provisioner timeout"))
+		time.Sleep(pollTimeoutSixMin)
+
+		ginkgo.By(fmt.Sprintf("Starting %v on the vCenter host", serviceName))
+		err = invokeVCenterServiceControl(startOperation, serviceName, vcAddress)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = false
+		err = waitVCenterServiceToBeInState(serviceName, vcAddress, startMessage)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		time.Sleep(totalResizeWaitPeriod)
+	}
+	//After service restart
+	bootstrap()
+
+	ginkgo.By("Waiting for file system resize to finish")
+	for _, claim := range pvclaims {
+		claims, err := waitForFSResize(claim, client)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		pvcConditions := claims.Status.Conditions
+		expectEqual(len(pvcConditions), 0, "pvc should not have conditions")
+	}
```

Comment:
Also check the file system after the expansion.

compare the filesystem size before and after the expansion

fsSize, err = getFSSizeMb(f, pod)
gomega.Expect(err).NotTo(gomega.HaveOccurred())		
gomega.Expect(fsSize).Should(gomega.BeNumerically(">", originalSizeInMb), fmt.Sprintf("error updating filesystem size for %q. Resulting filesystem size is %d", pvclaim.Name, fsSize))
		ginkgo.By("File system resize finished successfully")

Reply:
this is a good to have code - but not required for idempotency testing